src/Microsoft.Identity.Web.TokenAcquisition/TokenAcquisition.cs

```diff
@@ -585,10 +608,81 @@ private bool IsInvalidClientCertificateOrSignedAssertionError(MsalServiceExcepti
+            MergedOptions mergedOptions, ManagedIdentityOptions managedIdentityOptions)
+        {
+            string key = GetCacheKeyForManagedId(managedIdentityOptions);
+
+            // Check if the application is already built, if so return it without grabbing the key
+            if (_managedIdentityApplicationsByClientId.TryGetValue(key, out IManagedIdentityApplication? application) && application != null)
+            {
+                return application;
+            }
+
+            // Lock the potential write of the dictionary to prevent multiple threads from creating the same application.
+            await _managedIdSemaphore.WaitAsync();
+            try
+            {
+                // Check if the application is already built (could happen between previous check and obtaining the key)
+                if (_managedIdentityApplicationsByClientId.TryGetValue(key, out application) && application != null)
+                {
+                    return application;
+                }
+
+                // Set managedIdentityId to the correct value for either system or user assigned
+                ManagedIdentityId managedIdentityId;
+                if (key == SystemAssignedManagedIdentityKey)
+                {
+                    managedIdentityId = ManagedIdentityId.SystemAssigned;
+                }
+                else
+                {
+                    managedIdentityId = ManagedIdentityId.WithUserAssignedClientId(key);
+                }
+    
+                // Build the application
+                application = BuildManagedIdentityApplication(
+                    managedIdentityId,
+                    mergedOptions.ConfidentialClientApplicationOptions.EnablePiiLogging
+                );
 
+                // Add the application to the cache
+                _managedIdentityApplicationsByClientId.TryAdd(key, application);
+            } finally
+            {
+                // Now that the dictionary is updated, release the semaphore
+                _managedIdSemaphore.Release();
+            }
             return application;
         }
 
+        /// <summary>
+        /// Gets the key value for the Managed Identity cache, the default key for system-assigned identity is used if there is no
+        /// clientId for a user-assigned identity specified.
+        /// </summary>
+        /// <param name="managedIdOptions">Holds the clientId for managed identity if none is present</param>
+        /// <returns>A key value for the Managed Identity cache</returns>
+        internal /* for testing */ static string GetCacheKeyForManagedId(ManagedIdentityOptions managedIdOptions) 
+        {
+            if (managedIdOptions.UserAssignedClientId.IsNullOrEmpty())
+            {
+                return SystemAssignedManagedIdentityKey;
+            }
+            else
+            {
+                return managedIdOptions.UserAssignedClientId!;
```

Comment:
do we know this is not null?

Reply:
Since the if checks this value for NullOrEmpty, the else shouldn't have any way to be null. 

(Other than if someone messes with the object between the if check and dereference but that would mean they implemented incorrectly as this value should come from the appsettings.json config file.)

Reply:
maybe it's safer to use `string.IsNullOrEmpty(managedIdOptions.UserAssignedClientId)` and remove the `!`